manager/controllers/app/solve.go

```diff
@@ -59,10 +63,78 @@ func solve(env *datapath.Environment, datasets []datapath.DataInfo, log *zerolog
 }
 
 // perform basic checks before searching for a solution for a dataset
-func validateBasicConditions(env *datapath.Environment, log *zerolog.Logger) error {
+func validateBasicConditions(env *datapath.Environment, datasets []datapath.DataInfo, log *zerolog.Logger) error {
 	if len(env.Modules) == 0 {
 		log.Error().Msg(NoDeployedModules)
 		return errors.New(NoDeployedModules)
 	}
+	for i := range datasets {
+		dataset := &datasets[i]
+		if dataset.Context.Flow == "" || dataset.Context.Flow == taxonomy.ReadFlow {
+			if errorMessage := validateApplicationProtocol(env, dataset); errorMessage != "" {
+				log.Error().Str(managerUtils.FybrikAppUUID, dataset.Configuration.UUID).Msg(errorMessage)
+				return errors.New(errorMessage)
+			}
+			if errorMessage := validateAssetProtocol(env, dataset); errorMessage != "" {
+				log.Error().Str(managerUtils.FybrikAppUUID, dataset.Configuration.UUID).Msg(errorMessage)
+				return errors.New(errorMessage)
+			}
+		}
+	}
 	return nil
 }
+
+// create interface string to print in error messages
+func createInterfaceString(interfacePtr *taxonomy.Interface) string {
+	interfaceStr := string(interfacePtr.Protocol)
+	if interfacePtr.DataFormat != "" {
+		interfaceStr = interfaceStr + ", " + string(interfacePtr.DataFormat)
+	}
+	return interfaceStr
+}
+
+// check if any deployed module provides the requested read api by the application
+// return an empty string if such module exists, and an error message if not
+func validateApplicationProtocol(env *datapath.Environment, dataset *datapath.DataInfo) string {
```

Comment:
It is better to return error instead of string - nil for success and errors.New(`errorMessage`) for an error with `errorMessage`

Reply:
done